### ISSUE TITLE
`torch.profiler.record_function` doesn't register kernels from `backward` function

### ISSUE THREAD

**Issue by anmyachev** [pytorch/pytorch#144778]
### 🐛 Describe the bug

`__profile_kernel_of_func` (`record_function` label) shows zero timings for XPU (maybe for `CUDA` the situation is the same, but I have no way to check) unless `record_function` is used inside `backward` function.

```python
import torch
from torch.profiler import profile, ProfilerActivity, record_function

cache_size = 256
device = "xpu"

class _attention(torch.autograd.Function):

    @staticmethod
    def forward(ctx, cache):
        ctx.save_for_backward(cache)
        return cache

    @staticmethod
    def backward(ctx, triton_do):
        cache = ctx.saved_tensors
        # with record_function("__profile_kernel_of_func"): <- using this you can get the necessary timings
        cache[0].zero_()
        return cache
    

attention = _attention.apply

cache = torch.randn((128, 128), dtype=torch.float32, device=device, requires_grad=True)
triton_o = attention(cache)

triton_do = torch.randn_like(triton_o)
triton_fn = lambda: triton_o.backward(triton_do, retain_graph=True)

with profile(activities=[ProfilerActivity.CPU, ProfilerActivity.XPU]) as prof:
    with record_function("__profile_kernel_of_func"):
        triton_fn()
    torch.xpu.synchronize()

print(prof.events())

```

Output:
```

# case1 - `record_function` is not used in `backward` function and one can see that timings are zero

-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg      Self XPU    Self XPU %     XPU total  XPU time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                               __profile_kernel_of_func        90.70%     193.820ms        90.70%     193.820ms     193.820ms       0.000us         0.00%       0.000us       0.000us             1  
autograd::engine::evaluate_function: _attentionBackw...         0.01%      20.711us         3.38%       7.220ms       7.220ms       0.000us         0.00%       8.640us       8.640us             1  
                                     _attentionBackward         0.04%      86.246us         3.37%       7.199ms       7.199ms       0.000us         0.00%       8.640us       8.640us             1  
                                            aten::zero_         0.03%      53.549us         3.33%       7.113ms       7.113ms       0.000us         0.00%       8.640us       8.640us             1  
                                            aten::fill_         3.17%       6.772ms         3.30%       7.059ms       7.059ms       8.640us        50.00%       8.640us       8.640us             1  
                                  urEnqueueKernelLaunch         0.13%     287.035us         0.13%     287.035us     287.035us       0.000us         0.00%       0.000us       0.000us             1  
at::native::xpu::VectorizedElementwiseKernel<4, at::...         0.00%       0.000us         0.00%       0.000us       0.000us       8.640us        50.00%       8.640us       8.640us             1  
autograd::engine::evaluate_function: torch::autograd...         0.00%       5.280us         5.93%      12.663ms      12.663ms       0.000us         0.00%       8.640us       8.640us             1  
                        torch::autograd::AccumulateGrad         0.02%      34.944us         5.92%      12.658ms      12.658ms       0.000us         0.00%       8.640us       8.640us             1  
                                aten::new_empty_strided         0.00%       7.971us         0.01%      24.983us      24.983us       0.000us         0.00%       0.000us       0.000us             1  
                                    aten::empty_strided         0.01%      17.012us         0.01%      17.012us      17.012us       0.000us         0.00%       0.000us       0.000us             1  
                                            aten::copy_         5.86%      12.514ms         5.90%      12.598ms      12.598ms       8.640us        50.00%       8.640us       8.640us             1  
                                  urEnqueueKernelLaunch         0.04%      83.934us         0.04%      83.934us      83.934us       0.000us         0.00%       0.000us       0.000us             1  
at::native::xpu::VectorizedElementwiseKernel<4, at::...         0.00%       0.000us         0.00%       0.000us       0.000us       8.640us        50.00%       8.640us       8.640us             1  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 213.703ms
Self XPU time total: 17.280us


case #2 - `record_function` is used in `backward` function

-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg      Self XPU    Self XPU %     XPU total  XPU time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                               __profile_kernel_of_func        90.60%     218.883ms        90.60%     218.883ms     218.883ms       0.000us         0.00%       0.000us       0.000us             1  
autograd::engine::evaluate_function: _attentionBackw...         0.01%      20.057us         3.60%       8.690ms       8.690ms       0.000us         0.00%       8.320us       8.320us             1  
                                     _attentionBackward         0.03%      81.833us         3.59%       8.670ms       8.670ms       0.000us         0.00%       8.320us       8.320us             1  
                               __profile_kernel_of_func         0.10%     230.601us         3.55%       8.588ms       8.588ms       0.000us         0.00%       8.320us       8.320us             1  
                                            aten::zero_         0.02%      54.334us         3.46%       8.358ms       8.358ms       0.000us         0.00%       8.320us       8.320us             1  
                                            aten::fill_         3.30%       7.984ms         3.44%       8.304ms       8.304ms       8.320us        49.06%       8.320us       8.320us             1  
                                  urEnqueueKernelLaunch         0.13%     319.102us         0.13%     319.102us     319.102us       0.000us         0.00%       0.000us       0.000us             1  
at::native::xpu::VectorizedElementwiseKernel<4, at::...         0.00%       0.000us         0.00%       0.000us       0.000us       8.320us        49.06%       8.320us       8.320us             1  
autograd::engine::evaluate_function: torch::autograd...         0.00%       6.131us         5.80%      14.019ms      14.019ms       0.000us         0.00%       8.640us       8.640us             1  
                        torch::autograd::AccumulateGrad         0.02%      42.298us         5.80%      14.013ms      14.013ms       0.000us         0.00%       8.640us       8.640us             1  
                                aten::new_empty_strided         0.00%       9.754us         0.01%      26.166us      26.166us       0.000us         0.00%       0.000us       0.000us             1  
                                    aten::empty_strided         0.01%      16.412us         0.01%      16.412us      16.412us       0.000us         0.00%       0.000us       0.000us             1  
                                            aten::copy_         5.73%      13.855ms         5.77%      13.945ms      13.945ms       8.640us        50.94%       8.640us       8.640us             1  
                                  urEnqueueKernelLaunch         0.04%      90.206us         0.04%      90.206us      90.206us       0.000us         0.00%       0.000us       0.000us             1  
at::native::xpu::VectorizedElementwiseKernel<4, at::...         0.00%       0.000us         0.00%       0.000us       0.000us       8.640us        50.94%       8.640us       8.640us             1  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 241.593ms
Self XPU time total: 16.960us
```

### Versions

Pytorch pin: `1e881ceecfe80532206ca4e0acb64391fab8b935`.

**Comment by anmyachev** [comment]
@guangyey @Stonepia @etaf could you take a look? This may be the intended behavior, but I couldn't find anything about it in the profiler documentation.

**Comment by davidberard98** [comment]
This is intended behavior. It's probably possible to implement something that creates a corresponding backward record_function event (either in post-processing by following the fwd-bwd links) or by inserting an autograd node, but I don't think any work like this is planned.

**Comment by Stonepia** [comment]
The behavior for XPU is the same with CUDA:

```
# Case 1. You can find that the `__profile_kernel_of_func` also have 0 self CUDA time.
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                               __profile_kernel_of_func        95.04%     383.897ms        95.04%     383.897ms     383.897ms       0.000us         0.00%       0.000us       0.000us             1
autograd::engine::evaluate_function: _attentionBackw...         0.01%      30.810us         4.26%      17.211ms      17.211ms       0.000us         0.00%       1.920us       1.920us             1
                                     _attentionBackward         0.40%       1.611ms         4.25%      17.180ms      17.180ms       0.000us         0.00%       1.920us       1.920us             1
                                            aten::zero_         0.49%       1.968ms         3.85%      15.568ms      15.568ms       0.000us         0.00%       1.920us       1.920us             1
                                            aten::fill_         0.31%       1.240ms         3.37%      13.601ms      13.601ms       1.920us        46.51%       1.920us       1.920us             1
                                       cudaLaunchKernel         3.06%      12.361ms         3.06%      12.361ms      12.361ms       0.000us         0.00%       0.000us       0.000us             1
void at::native::vectorized_elementwise_kernel<4, at...         0.00%       0.000us         0.00%       0.000us       0.000us       1.920us        46.51%       1.920us       1.920us             1
autograd::engine::evaluate_function: torch::autograd...         0.00%       6.250us         0.69%       2.787ms       2.787ms       0.000us         0.00%       2.208us       2.208us             1
                        torch::autograd::AccumulateGrad         0.38%       1.548ms         0.69%       2.780ms       2.780ms       0.000us         0.00%       2.208us       2.208us             1
                                aten::new_empty_strided         0.01%      35.630us         0.02%      75.410us      75.410us       0.000us         0.00%       0.000us       0.000us             1
                                    aten::empty_strided         0.01%      39.780us         0.01%      39.780us      39.780us       0.000us         0.00%       0.000us       0.000us             1
                                            aten::copy_         0.27%       1.096ms         0.29%       1.157ms       1.157ms       2.208us        53.49%       2.208us       2.208us             1
                                        cudaMemcpyAsync         0.01%      60.250us         0.01%      60.250us      60.250us       0.000us         0.00%       0.000us       0.000us             1
                         Memcpy DtoD (Device -> Device)         0.00%       0.000us         0.00%       0.000us       0.000us       2.208us        53.49%       2.208us       2.208us             1
                                  cudaDeviceSynchronize         0.01%      34.000us         0.01%      34.000us      34.000us       0.000us         0.00%       0.000us       0.000us             1
                                  cudaDeviceSynchronize         0.00%       3.990us         0.00%       3.990us       3.990us       0.000us         0.00%       0.000us       0.000us             1
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 403.932ms
Self CUDA time total: 4.128us
```
I think something need to be changed for the `torch.profiler.profile`. If I switch the profiler to ` torch.autograd.profiler.profile` and do profiling like below:

```Python
with torch.autograd.profiler.profile(use_device='cuda') as prof:
    with torch.autograd.profiler.record_function("__profile_kernel_of_func"):
        triton_fn()
    torch.cuda.synchronize()

print(prof.key_averages().table())
```
I can get the expected result, now the `__profile_kernel_of_func`'s CUDA time is not 0.

```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
                                        cudaEventRecord         0.07%     170.990us         0.07%     170.990us       8.549us       0.000us         0.00%       0.000us       0.000us            20
                               __profile_kernel_of_func        96.19%     219.900ms        96.19%     219.900ms     219.900ms     219.962ms        96.24%     219.962ms     219.962ms             1
autograd::engine::evaluate_function: _attentionBackw...         0.01%      30.910us         3.63%       8.305ms       8.305ms      51.000us         0.02%       8.318ms       8.318ms             1
                                     _attentionBackward         0.13%     307.360us         3.61%       8.263ms       8.263ms     313.000us         0.14%       8.267ms       8.267ms             1
                                            aten::zero_         0.05%     118.700us         3.48%       7.950ms       7.950ms     123.000us         0.05%       7.954ms       7.954ms             1
                                            aten::fill_         0.03%      72.430us         3.42%       7.821ms       7.821ms       7.831ms         3.43%       7.831ms       7.831ms             1
                                       cudaLaunchKernel         3.39%       7.748ms         3.39%       7.748ms       7.748ms       0.000us         0.00%       0.000us       0.000us             1
autograd::engine::evaluate_function: torch::autograd...         0.00%       8.200us         0.11%     262.170us     262.170us      16.000us         0.01%     264.000us     264.000us             1
                        torch::autograd::AccumulateGrad         0.02%      56.310us         0.11%     244.430us     244.430us      61.000us         0.03%     248.000us     248.000us             1
                                aten::new_empty_strided         0.01%      28.690us         0.04%      90.020us      90.020us      37.000us         0.02%      92.000us      92.000us             1
                                    aten::empty_strided         0.02%      52.470us         0.02%      52.470us      52.470us      55.000us         0.02%      55.000us      55.000us             1
                                            aten::copy_         0.02%      39.600us         0.04%      91.580us      91.580us      95.000us         0.04%      95.000us      95.000us             1
                                        cudaMemcpyAsync         0.02%      51.980us         0.02%      51.980us      51.980us       0.000us         0.00%       0.000us       0.000us             1
                                  cudaDeviceSynchronize         0.01%      29.960us         0.01%      29.960us      14.980us       0.000us         0.00%       0.000us       0.000us             2
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
Self CPU time total: 228.615ms
Self CUDA time total: 228.544ms
```

@zejun-chen do you have any ideas?
